### PR TITLE
download api added

### PR DESCRIPTION
**Pull request by rakeshSgr**

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -242,3 +242,5 @@ AP_USERS_ORGANISATION_ID = ""
 CASSANDRA_HOST = ""
 CASSANDRA_PORT = ""
 CASSANDRA_DB = ""
+
+CLOUD_PATH = ""
```
Comment (aks30) [inline .env.sample]: @rakeshSgr - No need to provide folder via env. Use existing uploads folder

File: controllers/v1/cloud-services/aws.js [hunk 2]
```diff
@@ -72,6 +72,9 @@ module.exports = class Aws {
 
             try {
 
+                console.log(" request ---", req.body.filePaths ,"--",
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js]: @rakeshSgr - Please remove console logs

File: controllers/v1/cloud-services/aws.js [hunk 3]
```diff
@@ -179,5 +182,80 @@ module.exports = class Aws {
...
+      * To upload file cloud 
+      * @method
+      * @name upload
+      * @param  {Request}  req  request body.
+      * @param  {files}  req.files.file -actual file to upload
+      * @param  {String}  req.body.presignedUrl -cloud url to upload
+      * @param  {String}  req.body.fileName - name of file
+      * @returns {JSON} Response with status and message.
+    */
+   async uploadToPresignedUrl(req) {
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js]: @rakeshSgr - Create one upload endpoint that internally uploads it to google cloud storage. Also make it secure by internal access token

File: controllers/v1/cloud-services/aws.js [hunk 4]
```diff
@@ -179,5 +179,100 @@ module.exports = class Aws {
...
+     * "bucketName":"sl-dev-storage",
+     * "storage": "GC"
+     *  
+     * }
+     * @apiSampleRequest /kendra/api/v1/cloud-services/aws/uploadFile
+     * @apiUse successBody
+     * @apiUse errorBody
+     * @apiParamExample {json} Response:
+     * {
+     * "status": 200,
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js:201]: @rakeshSgr - Add message as well

File: .env.sample [hunk 5]
```diff
@@ -242,3 +242,5 @@ AP_USERS_ORGANISATION_ID = ""
 CASSANDRA_HOST = ""
 CASSANDRA_PORT = ""
 CASSANDRA_DB = ""
+
+UPLOAD_PATH = ""
```
Comment (aks30) [inline .env.sample]: @rakeshSgr - No need in env, put it in a file as a constant file

File: controllers/v1/cloud-services/aws.js [hunk 6]
```diff
@@ -179,5 +179,100 @@ module.exports = class Aws {
...
+            if(req.files && req.body.filePath && req.body.bucketName){
+                let response  = await filesHelpers.uploadFile(
+                    req.files.file,
+                    req.body.filePath,
+                    req.body.bucketName,
+                    constants.common.AWS_SERVICE
+            );
+            return resolve(response);
+
+            }else{
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js]: @rakeshSgr - Fix indentation and spacing

File: generics/middleware/authenticator.js [hunk 7]
```diff
@@ -98,7 +98,7 @@ module.exports = async function (req, res, next) {
     return
   }
 
-  if (req.path.includes("keywords")) {
+  if (req.path.includes("keywords") || req.path.includes("uploadFile1")) {
```
Comment (aks30) [inline generics/middleware/authenticator.js]: @rakeshSgr - What is uploadFile1 ? aws/uploadFile, azureUploadFile needs to be there

File: controllers/v1/cloud-services/azure.js [hunk 8]
```diff
@@ -241,6 +241,95 @@ module.exports = class Azure {
...
+      * @param  {files}  req.files.file -actual file to upload
+      * @param  {String}  req.body.bucketName - bucket name
+      * @param  {String}  req.body.filePath - file path of where to store
+      * @returns {JSON} Response with status and message.
+    */
+   async uploadFile(req) {
+    return new Promise(async (resolve, reject) => {
+
+        try {
+            if(req.files && req.body.filePath && req.body.bucketName){
```
Comment (aks30) [inline controllers/v1/cloud-services/azure.js]: @rakeshSgr - Please add indentation and space as per coding format.

File: controllers/v1/cloud-services/gcp.js [hunk 9]
```diff
@@ -180,5 +180,96 @@ module.exports = class Gcp {
     })
    }
 
+
+       /**
+     * @api {post} /kendra/api/v1/cloud-services/gcp/uploadFile  
+     * upload file to presignedUrl
```
Comment (aks30) [inline controllers/v1/cloud-services/gcp.js]: @rakeshSgr - Make upload as upload file to presignedUrl, => File upload to GCP

File: controllers/v1/cloud-services/gcp.js [hunk 10]
```diff
@@ -180,5 +180,96 @@ module.exports = class Gcp {
...
+     *   "etag": "CJ2tsf35xukCEAE=",
+     *   "timeCreated": "2020-05-22T07:31:55.836Z",
+     *   "updated": "2020-05-22T07:31:55.836Z",
+     *   "timeStorageClassUpdated": "2020-05-22T07:31:55.836Z"
+     *    }
+     * }
+     */
+
+    /**
+      * To upload file cloud
```
Comment (aks30) [inline controllers/v1/cloud-services/gcp.js]: @rakeshSgr - Change name to - Upload file to GCP

File: generics/middleware/authenticator.js [hunk 11]
```diff
@@ -98,7 +98,8 @@ module.exports = async function (req, res, next) {
     return
   }
 
-  if (req.path.includes("keywords")) {
+  if (req.path.includes("keywords") || req.path.includes("/aws/uploadFile") || 
+  req.path.includes("/gcp/uploadFile") || req.path.includes("/azure/uploadFile")) {
```
Comment (aks30) [inline generics/middleware/authenticator.js]: @rakeshSgr - Make this as an array and search in that.

File: module/files/helper.js [hunk 12]
```diff
@@ -7,404 +7,444 @@
  */
 
 // Dependencies
-const Zip = require("adm-zip");
-const fs = require("fs");
-const awsServices = require(ROOT_PATH + "/generics/services/aws");
-const googleCloudServices = require(ROOT_PATH + "/generics/services/google-cloud");
-const azureService = require(ROOT_PATH + "/generics/services/azure");
+const Zip = require('adm-zip')
```
Comment (aks30) [inline module/files/helper.js]: @rakeshSgr - Add semi-colons at the end of line

File: module/files/helper.js [hunk 13]
```diff
@@ -7,404 +7,444 @@
...
+   * Upload file in different services based on cloud storage provided.
+   * @method
+   * @name upload
+   * @param  {file}  - file to upload.
+   * @param  {filePathForBucket}  - file path where the file should upload.
+   * @param {String} - bucketName
+   * @returns {json} Response consists of links of uploaded file.
+   */
 
+  static uploadFile (file, filePathForBucket, bucketName, storage) {
```
Comment (aks30) [inline module/files/helper.js]: @rakeshSgr - Let's have a common function for upload. Also, fix indentation so no other changes come

File: .env.sample (hunk 1 above)
Comment (rakeshSgr) [reply .env.sample]: resolved

File: controllers/v1/cloud-services/aws.js (hunk 2 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/aws.js]: resolved

File: controllers/v1/cloud-services/aws.js (hunk 3 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/aws.js]: resolved

File: .env.sample (hunk 5 above)
Comment (rakeshSgr) [reply .env.sample]: resolved

File: controllers/v1/cloud-services/aws.js (hunk 6 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/aws.js]: resolved

File: generics/middleware/authenticator.js (hunk 7 above)
Comment (rakeshSgr) [reply generics/middleware/authenticator.js]: resolved

File: controllers/v1/cloud-services/azure.js (hunk 8 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/azure.js]: resolved

File: controllers/v1/cloud-services/gcp.js (hunk 9 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/gcp.js]: resolved

File: controllers/v1/cloud-services/gcp.js (hunk 10 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/gcp.js]: resolved

File: generics/middleware/authenticator.js (hunk 11 above)
Comment (rakeshSgr) [reply generics/middleware/authenticator.js]: resolved

File: module/files/helper.js (hunk 12 above)
Comment (rakeshSgr) [reply module/files/helper.js]: resolved

File: module/files/helper.js (hunk 13 above)
Comment (rakeshSgr) [reply module/files/helper.js]: resolved

File: controllers/v1/cloud-services/aws.js [hunk 14]
```diff
@@ -179,5 +180,97 @@ module.exports = class Aws {
...
+            } else {
+                return reject({
+                    status:
+                        httpStatusCode["bad_request"].status,
+                    message:httpStatusCode["bad_request"].message
+
+                });
+            }
+        } catch (error) {
+            console.log(error);
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js]: @rakeshSgr - No console logs please

File: controllers/v1/cloud-services/aws.js [hunk 15]
```diff
@@ -179,5 +180,97 @@ module.exports = class Aws {
...
+      * @param  {files}  req.files.file -actual file to upload
+      * @param  {String}  req.body.bucketName - bucket name
+      * @param  {String}  req.body.filePath - file path of where to store
+      * @returns {JSON} Response with status and message.
+    */
+   async uploadFile(req) {
+    return new Promise(async (resolve, reject) => {
+
+        try {
+            if (req.files && req.body.filePath && req.body.bucketName){
```
Comment (aks30) [inline controllers/v1/cloud-services/aws.js]: @rakeshSgr - Add a space before the curly brace

File: controllers/v1/cloud-services/azure.js [hunk 16]
```diff
@@ -241,6 +241,95 @@ module.exports = class Azure {
...
+      * @param  {files}  req.files.file -actual file to upload
+      * @param  {String}  req.body.bucketName - bucket name
+      * @param  {String}  req.body.filePath - file path of where to store
+      * @returns {JSON} Response with status and message.
+    */
+   async uploadFile(req) {
+    return new Promise(async (resolve, reject) => {
+
+        try {
+            if (req.files && req.body.filePath && req.body.bucketName){
```
Comment (aks30) [inline controllers/v1/cloud-services/azure.js]: @rakeshSgr - Add a space before the curly brace

File: controllers/v1/cloud-services/gcp.js [hunk 17]
```diff
@@ -180,5 +180,94 @@ module.exports = class Gcp {
...
+      * @param  {Request}  req  request body.
+      * @param  {files}  req.files.file -actual file to upload
+      * @param  {String}  req.body.bucketName - bucket name
+      * @param  {String}  req.body.filePath - file path of where to store
+      * @returns {JSON} Response with status and message.
+    */
+   async uploadFile(req) {
+    return new Promise(async (resolve, reject) => {
+        try {
+            if (req.files && req.body.filePath && req.body.bucketName){
```
Comment (aks30) [inline controllers/v1/cloud-services/gcp.js]: @rakeshSgr - Add a space before the curly brace

File: generics/middleware/authenticator.js [hunk 18]
```diff
@@ -93,13 +93,20 @@ module.exports = async function (req, res, next) {
...
-  if ( req.path.includes("bodh/search") || req.path.includes("bodh/request") ) {
+  if (req.path.includes("bodh/search") || req.path.includes("bodh/request")) {
     next();
     return
   }
 
-  if (req.path.includes("keywords")) {
-    if(req.headers["internal-access-token"] !== process.env.INTERNAL_ACCESS_TOKEN) {
+  let internalAccessApiPaths = ["keywords", "/aws/uploadFile", "/gcp/uploadFile", "/azure/uploadFile"];
+  let found = false;
```
Comment (aks30) [inline generics/middleware/authenticator.js]: @rakeshSgr - Please use variable "**performInternalAccessTokenCheck**"

Comment (rakeshSgr) [reply generics/middleware/authenticator.js]: resolved

File: controllers/v1/cloud-services/gcp.js (hunk 17 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/gcp.js]: resolved

File: controllers/v1/cloud-services/azure.js (hunk 16 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/azure.js]: resolved

File: controllers/v1/cloud-services/aws.js (hunk 15 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/aws.js]: resolved

File: controllers/v1/cloud-services/aws.js (hunk 14 above)
Comment (rakeshSgr) [reply controllers/v1/cloud-services/aws.js]: resolved

**Pull request merged**